foundation-model-stack/fms-hf-tuning
# feat: DataProcessor v1

### Description of the change

This PR is to be merged after https://github.com/foundation-model-stack/fms-hf-tuning/pull/398 as it follows the changes. Once https://github.com/foundation-model-stack/fms-hf-tuning/pull/398 is merged this PR will be rebased on top of it.

This PR is to provide a data preprocessor framework which will enable flexibility to easily add more data preprocessing features in the future. This PR covers the following:

- Change base framework to the new configurable framework
   - Preprocessing utils function changed from `format_dataset` to `process_dataargs`
- Handle currently supported features using the new base framework
- Cleanup redundant code and port unit tests.
- Unit testing data preprocessor
- Unit testing setup data preprocessor

This PR does not explicitly enable any new features in fms-hf-tuning, it's purpose is to more easily allow new features to be added in the future by changing the backend.

Co-Authored by @willmj @Abhishek-TAMU 

### How to verify the PR

- New/refactored unit tests located in `testing/data`
- Unit tests of `train()` function
- Run e2e tuning + vLLM (see example [build](https://v3.travis.ibm.com/github/ai-foundation/fmaas-tuning-docs/builds/30211892))

### Was the PR tested
